File: Day Challenge/20230624 Longest Arithmetic Subsequence/1027. Longest Arithmetic Subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestArithSeqLength = function (nums) {
  let dp = [];
  let max = 0;
  for (let i = 0; i < nums.length; i++) {
    dp.push({});
    for (let j = 0; j < i; j++) {
      let diff = nums[i] - nums[j];
      if (dp[j][diff]) {
        dp[i][diff] = dp[j][diff] + 1;
      } else {
        dp[i][diff] = 2;
      }
      max = Math.max(max, dp[i][diff]);
    }
  }
  return max;
};

File: Day Challenge/20230624 Longest Arithmetic Subsequence/1027. Longest Arithmetic Subsequence.py
class Solution(object):
    def longestArithSeqLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        diffSet = set()
        answer = 0
        logestLst = []
        for i in range(len(nums)):
            curr = nums[i]
            for j in range(i+1, len(nums)):
                next = nums[j]
                diff = next - curr
                if diff in diffSet:
                    continue
                else:
                    diffSet.add(diff)
                pt1, pt2 = j, j+1
                steps = 2
                lst = [nums[i], nums[j]]
                while pt2 < len(nums):
                    if nums[pt2] - nums[pt1] == diff:
                        steps += 1
                        lst.append(nums[pt2])
                        pt1 = pt2
                        pt2 += 1
                    else:
                        pt2 += 1
                if len(lst) > len(logestLst):
                    logestLst = list(lst)
                answer = max(steps, answer)

        return answer
